ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/transform/KsValueTransformer.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2022 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.streams.transform;
+
+import static java.util.Objects.requireNonNull;
+
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.execution.transform.KsqlProcessingContext;
+import io.confluent.ksql.execution.transform.KsqlTransformer;
+import java.util.Optional;
+import org.apache.kafka.streams.kstream.ValueTransformerWithKey;
+import org.apache.kafka.streams.processor.ProcessorContext;
+
+/**
+ * A Kafka-streams transformer
+ *
+ * <p>Maps a implementation agnostic {@link KsqlTransformer} to a implementation specific {@link
+ * ValueTransformerWithKey}.
+ *
+ * @param <K> the type of the key
+ * @param <R> the return type
+ */
+public class KsValueTransformer<K, R> implements ValueTransformerWithKey<K, GenericRow, R> {
+
+  private final KsqlTransformer<K, R> delegate;
+  private Optional<KsProcessingContext> context;
+
+  public KsValueTransformer(final KsqlTransformer<K, R> delegate) {
+    this.delegate = requireNonNull(delegate, "delegate");
+    this.context = Optional.empty();
```

Comment:
If that's sane, I wonder if it'd be cleaner to have just one class that transforms/carries (KInT, GenericRow) -> (KOutT, R).

Different constructors or static method could build appropriate instances of the class?

(This is a weakly held opinion / thought.  Disregard readily.:))

Reply:
Technically yes, but as mentioned above, it might trigger an (unnecessary) repartition step downstream is we don't change the key. -- KS runtime does not know what change is applied and need to conservatively assume that the key was changed if `Transformer` is used, and it can only rely on a unmodified key if `ValueTransformer` is used.